Support for AIX 7.1 TL4.

johnnys66 opened the pull request:
Hi Jakub,

I finally got started with IPA on AIX. It's working with just a few changes.

- Added function getgrset in get_user_groups to query the users groups, funcion getgrent in AIX does not query ldap, only nis and local.

Commit:
Support for AIX 7.1 TL4.

Added function getgrset in get_user_groups to query the users groups, funcion getgrent in AIX does not query ldap, only nis and local.

Resolves:
https://github.com/jhrozek/pam_hbac/issues/74

jhrozek commented:
I'm sorry I was silent about this PR for a whole month, but I was busy with work...

I'm really glad you found the time to contribute some more to pam_hbac! I have some comments and some questions, see the review of the PR. Please let me know if I should provide some patches you can squash in your branch to save you time.

jhrozek commented on doc/README.AIX:
```diff
@@ -0,0 +1,64 @@
+
+This build does not have manpages. In AIX the utility a2x does something other than
+converting asciidoc. It must be removed from the PATH variable otherwise the build
+will fail. 
```
Hmm, this is unfortunate, do you think it would be a good idea to provide a configure-time switch `--without-manpages`` so that admins on AIX could just skip building the man pages altogether?

johnnys66 replied:
Yes, totally agree.

jhrozek commented on src/pam_hbac_obj.c:
```diff
@@ -172,6 +172,27 @@ get_user_groups(const char *name, gid_t primary_gid,
+    max_group_len = sysconf(_SC_LOGIN_NAME_MAX);
+
+    /* string containing comma separated list of gids the user belongs to */
+    gid_list_s = malloc(sizeof(char)*max_group_len*NGROUPS_MAX);
```
Hmm, here are two lines that both assign to gid_list_s and both allocate memory, so the memory that comes from malloc is lost.. At least some documentation I found (https://www.ibm.com/support/knowledgecenter/ssw_aix_71/com.ibm.aix.basetrf1/getgrset.htm) seems to suggest that `getgrset` already returns an allocated list, therefore I think the malloc andthe max_group_len variable are not needed?

johnnys66 replied:
I think you are right, no need for the malloc and the variable. My C programming skills are not the best :).

jhrozek commented on src/pam_hbac_obj.c:
```diff
@@ -172,6 +172,27 @@ get_user_groups(const char *name, gid_t primary_gid,
+
+    /* string containing comma separated list of gids the user belongs to */
+    gid_list_s = malloc(sizeof(char)*max_group_len*NGROUPS_MAX);
+    gid_list_s = getgrset(name);
```
The memory pointed to by the gid_list_s pointer is lost, I think `free()` needs to be called just before `#else`.

johnnys66 replied:
The link to the documentation you posted says that it should be freed yes.

jhrozek commented on src/pam_hbac_obj.c:
```diff
@@ -172,6 +172,27 @@ get_user_groups(const char *name, gid_t primary_gid,
+    gid_list_s = malloc(sizeof(char)*max_group_len*NGROUPS_MAX);
+    gid_list_s = getgrset(name);
+
+    gid_s = malloc(sizeof(gid_t));
```
This chunk of memory is not freed either

jhrozek reviewed (changes requested):
Please let me know if I should provide patches to squash in or the patch to conditionalize man pages generation.

johnnys66 commented:
I think all your suggestions make sense. If you can provide the patches I'll be happy to test them.

jhrozek committed:
BUILD: Add an --enable-man-pages configuration option

Resolves:
https://github.com/jhrozek/pam_hbac/issues/100

Adds a new config-time option --enable-man-pages. Man pages are enabled
by default, but when this option is set to --disable-man-pages, even
though a2x might be detected on the system, man pages will not be
generated.

jhrozek commented:
OK, I pushed your patch along with one of mine that implements the option to avoid building man pages if you run `configure` with `--disable-man-pages` and another that I would like to squash into yours that mostly just checks allocation results and so on. But I didn't test the AIX codepath at all to be honest...but now I do have some access (albeit non-root) to an AIX machine, so hopefully I'll at least check out the groups-for-user branch.

jhrozek commented:
And I forgot to say I pushed the new branch here: https://github.com/jhrozek/pam_hbac/tree/aix
